From c4b15ddcc19dbb9b02cf60b7a7518e5af10f8f7d Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Tor=20Arne=20Vestb=C3=B8?= <tor.arne.vestbo@qt.io>
Date: Wed, 1 Oct 2025 15:04:50 +0200
Subject: [PATCH 375/553] Don't override NSImage aspect ratio in [NSImage
 imageFromQIcon]

The size is input to the icon engine, but if the engine then produces
icons of a different size, and different aspect ratio in particular,
we can't naively apply the input size.

Doing so would manifest as skewed icons in menus and window title
bars.

Pick-to: 6.8 6.5
Change-Id: Ifd055e46cc268aec98f2157036d012b7d70d752a
Reviewed-by: Volker Hilsheimer <volker.hilsheimer@qt.io>
(cherry picked from commit 8a6a566489b55b185f998b72adafee1e3f204893)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 src/gui/painting/qcoregraphics.mm | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/src/gui/painting/qcoregraphics.mm b/src/gui/painting/qcoregraphics.mm
index 6e57085b50f..6bc41658a56 100644
--- a/src/gui/painting/qcoregraphics.mm
+++ b/src/gui/painting/qcoregraphics.mm
@@ -162,8 +162,10 @@ QT_END_NAMESPACE
 
     [nsImage setTemplate:icon.isMask()];
 
-    if (!size.isNull())
-        nsImage.size = QSizeF(size).toCGSize();
+    if (!size.isNull()) {
+        auto imageSize = QSizeF::fromCGSize(nsImage.size);
+        nsImage.size = imageSize.scaled(size, Qt::KeepAspectRatio).toCGSize();
+    }
 
     return nsImage;
 }
-- 
2.51.2

